src/NodeApi/JSInvalidScopeException.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using Microsoft.JavaScript.NodeApi.Interop;
+
+namespace Microsoft.JavaScript.NodeApi;
+
+/// <summary>
+/// An exception that was caused by an attempt to access a JavaScript value without any
+/// <see cref="JSValueScope" /> established on the current thread, or from a thread associated
+/// with a different environment / root scope.
+/// </summary>
+/// <remarks>
+/// All JavaScript values are created within a scope that is bound to the thread that runs the
+/// JS environment. They can only be accessed from the same thread and only as long as the scope
+/// is still valid (not disposed).
+/// </remarks>
+/// <seealso cref="JSSynchronizationContext"/>
+public class JSInvalidScopeException : InvalidOperationException
+{
+    /// <summary>
+    /// Creates a new instance of <see cref="JSInvalidScopeException" /> with a 
+    /// current scope and message.
+    /// </summary>
+    public JSInvalidScopeException(
+        JSValueScope? currentScope,
```

Comment:
>JSValueScope? currentScope [](http://example.com/codeflow?start=8&length=26)

Since it is always must be `JSValueScope.Current`, I wonder if we should simplify the constructor and avoid passing it as a parameter, and initialize it from `JSValueScope.Current` in the constructor.

Reply:
No, because if the current scope is null then `JSValueScope.Current` will throw this exception and that would be infinite recursion.

I'm not sure what happened with the diff: `JSInvalidScopeException.cs` was supposed to be deleted because it was renamed to `JSInvalidThreadAccessException` at your suggestion.

Reply:
I was surprised to see two identical classes with different names. :)

Reply:
I deleted `JSInvalidScopeException.cs` in an update.